src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -145,8 +142,36 @@ object VSMSubgen {
     tGen = Genotype.genDosage, isDosage = true)
 }
 
-class VariantSampleMatrix[T](val hc: HailContext, val metadata: VariantMetadata,
-  val rdd: OrderedRDD[Locus, Variant, (Annotation, Iterable[T])])(implicit tct: ClassTag[T]) extends JoinAnnotator {
+class VariantSampleMatrix[T](val hc: HailContext, val metadata: VSMMetadata,
+  val ast: MatrixAST[T])(implicit val tct: ClassTag[T]) extends JoinAnnotator {
+
+  def this(hc: HailContext,
+    metadata: VSMMetadata,
+    localValue: VSMLocalValue,
+    rdd: OrderedRDD[Locus, Variant, (Annotation, Iterable[T])])(implicit tct: ClassTag[T]) =
+    this(hc, metadata,
+      MatrixLiteral(
+        MatrixType(metadata),
+        MatrixValue(localValue, rdd)))
+
+  def this(hc: HailContext, fileMetadata: VSMFileMetadata,
+    rdd: OrderedRDD[Locus, Variant, (Annotation, Iterable[T])])(implicit tct: ClassTag[T]) =
+    this(hc, fileMetadata.metadata, fileMetadata.localValue, rdd)
+
+  lazy val value = {
+    val opt = MatrixAST.optimize(ast)
+    println("ast", ast)
+    println("opt", opt)
+    opt.execute(hc)
+  }
+
+  def globalAnnotation: Annotation = value.localValue.globalAnnotation
+
+  def sampleIds: IndexedSeq[String] = value.localValue.sampleIds
+
+  def sampleAnnotations: IndexedSeq[Annotation] = value.localValue.sampleAnnotations
```

Comment:
I do hate duplicating all these functions, but I don't think this works because imports are not visible outside the class:

```
class B(val i: Int)

class A {
  lazy val b: B = {
    println("make b");
    new B(5)
  }
  import b._
}

scala> val a = new A()
a: A = A@71318ec4

scala> a.i
<console>:15: error: value i is not a member of A
       a.i
         ^
```